python/lsst/pipe/base/graph.py

```diff
@@ -96,18 +135,67 @@ def __init__(self, iterable=None):
     def quanta(self):
         """Iterator over quanta in a graph.
 
+        Quanta are returned in unspecified order.
+
         Yields
         ------
         taskDef : `TaskDef`
             Task definition for a Quantum.
-        quantum : `Quantum`
+        quantum : `~lsst.daf.butler.Quantum`
             Single quantum.
         """
         for taskNodes in self:
             taskDef = taskNodes.taskDef
             for quantum in taskNodes.quanta:
                 yield taskDef, quantum
 
+    def orderedTaskNodes(self):
+        """Return topologically ordered task nodes.
+
+        Yields
+        ------
+        nodes : `QuantumGraphTaskNodes`
+        """
+        # Tasks in a graph are probably topologically sorted already but there
+        # is no guarantee for that. Just re-construct Pipeline and order tasks
+        # in a pipeline using existing method.
+        nodesMap = {id(item.taskDef): item for item in self}
+        pipeline = orderPipeline(Pipeline(item.taskDef for item in self))
+        for taskDef in pipeline:
+            yield nodesMap[id(taskDef)]
+
+    def orderedQuanta(self):
+        """Return topologically ordered Quanta and their pre-requisites.
+
+        This method enumerates all Quanta in a graph and iterates over
+        them in topological order so that pre-requisites for given Quantum
+        are always returned before the Quantum.
+
+        Yields
+        ------
+        quantumData : `QuantumIterData`
+        """
+        index = 0
+        outputs = {}  # maps DataId to its producing quantum index
```

Comment:
I don't think this is safe: an output needs to be keyed by DatasetType as well as DataId.  If `DatasetRef` were equality comparable and hashable, you could just use that, but I think it's tricky to implement because it's unclear whether it should compare `dataset_id` or just DatasetType and DataId.  After DM-16227 you could use the `DatasetRef.hash()` I've added there as the key, but maybe it'd be better to just use a tuple of `(ref.dataset_type_name, ref.dataId)` here anyway.

Reply:
😶 Somehow I keep forgetting about DatasetType, thanks for spotting this. Will fix with a tuple as you suggest.